File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        sums,flag = 0,False
        def has(root):
            nonlocal sums,flag
            if root == None:
                return
            sums += root.val
            has(root.left)
            has(root.right)
            if(root.left == None and root.right == None and sums == targetSum):
                flag = True
            sums -= root.val
        has(root)
        return flag

File: 112-path-sum/path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean traverse(TreeNode root,int pathSum,int targetSum){
    if(root == null){
        return false;
    }
    pathSum = pathSum + root.val;
    if((root.left == null) && (root.right == null))
        return pathSum == targetSum;
    return traverse(root.left,pathSum,targetSum) || traverse(root.right,pathSum,targetSum);
    }
    public boolean hasPathSum(TreeNode root, int targetSum) 
    {
        if(root != null)
          return traverse(root,0,targetSum);
        return false;
    }
}